feat(docker): Introduce reworked docker setup with jar

MichaelsJP opened the pull request:
This PR introduces a completely overhauled docker setup. Due to the nature of the changes, the old docker setup isn't compatible with this one. No new major incompatibilities are to be expected soon.

By moving away from the old (almost 6-year-old docker structure) we now have the following features available:
- Coherent user permissions
- Clear mount points
- Well documented docker-compose.yml
- Reworked entrypoint with log levels, sensible log messages, debug mode config less configuration etc.
- Reworked dockerfile that is roughly 1/3 in terms of code lines and a lot less complex.
- We now use the jar file in the container. No more .war-File back and forth (finally).
- Multiple ways of configuring ors: config files (yml, deprecated json), container env variables á la `ors.*.someting`
- **Cloud-ready image**. Te image can be 100% configured by environment variables (env files) and therefore run in any cloud and Kubernetes infrastructure without much hustle.
- The image itself can be used as a base for custom cloud images as well.

### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box
when you have checked you have done them): -->
- [ ] 1. I have [**rebased**][rebase] the latest version of the main branch into my feature branch and all conflicts
         have been resolved.
- [ ] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [ ] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [ ] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [ ] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes # .

### Information about the changes
- Key functionality added:
- Reason for change:

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
-

### Required changes to ors config (if applicable)
-

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/main/CONTRIBUTE.md#pull-request-guidelines

MichaelsJP committed:
feat(docker): Add the docker setup with jar

This introduces a completely revamped docker setup. The missing parts are the workflows and the docker-compose.yml.

Without any configuration the container uses an example yml config shipped with it.
Graph locations are parsed from envs, json and yml files

- Introduce container log levels and add adjustments
- Upgrade build image to java 21 (still compiled to 17)

MichaelsJP committed:
chore(config-conversion): Rename the parser script

MichaelsJP committed:
ci(workflow): Update used actions version

MichaelsJP committed:
ci(workflow): Update the docker build & test to the new setup

- use ors default port in container
- Add a basic docker compose check for built image

MichaelsJP committed:
feat(docker): Migrate BUILD_GRAPHS to REBUILD_GRAPHS

This has been a misleading variable for some time. This introduces REBUILD_GRAPHS and keeps BUILD_GRAPHS in compatibility for now.

MichaelsJP committed:
feat(docker): Enhance the documentation in the compose file

MichaelsJP committed:
fix(docker): Add more user related optimizations

The setup now detects if the user runs the container with a less privileged user and prints out necessary data for that

- improve container handling
- add default java opts & information to docker-compose environment

Co-authored-by: Amandus Butzer <amandus.butzer@heigit.org>

TheGreatRefrigerator commented:
rebased on main

TheGreatRefrigerator committed:
chore: ignore generated folders

TheGreatRefrigerator commented:
rebased on main and fixed conflicts

TheGreatRefrigerator committed:
chore(config-conversion): rework & add properties conversion

- move application.yml validation to separate file
- add error exit codes
- use yq to edit & validate
- use sed instead of awk (for os compatibility)
- move comments to echo output

- add a new script to parse application.yml to a properties file

The scripts are using yq to make the rewrite process more robust.
The less we use sed magic, the better.

TheGreatRefrigerator committed:
docs: adjust link to config parameters

TheGreatRefrigerator committed:
chore(config): automatic conversion of application.yml to ors-config.yml and ors-config.env

TheGreatRefrigerator commented:
Currenlty includes the nightly build commit
- Restructure commits
- rework conversion & extract validation
- resolve todos (link to docu & add default env properties for java opts)

TheGreatRefrigerator committed:
docs: adjust docker guide

TheGreatRefrigerator commented:
rebased on main

TheGreatRefrigerator committed:
docs: fix port in examples

TheGreatRefrigerator committed:
ci: adjust folder paths in docker-build-and-test.yml

TheGreatRefrigerator commented:
The test runs still fail due to the new setup prefering example-ors-config.yml over ors-config.yml.

Here the docs from my previous investigation
[docker_run_and_compose.md](https://github.com/GIScience/openrouteservice/files/14497520/docker_run_and_compose.md)

TheGreatRefrigerator committed:
refactor: change config path in docker

as ors is reading the config path from $HOME, this did not work
when running ors in a container as root.
We moved the config ors is using in the container to the default
lookup location '/etc/openrouteservice'.

introduces:
- copying example config to ors-config.yml if it doesn't exist
- not using example-config.yml by default
- adjust docker-compose.yml to changes

Co-authored-by: Jochen Haeussler <jochen.haeussler@heigit.org>

TheGreatRefrigerator committed:
refactor: use example-heidelberg.osm.gz

switch to usage of copied file in /home/ors/files instead of root folder

Co-authored-by: Julian Psotta <julian.psotta@heigit.org>
Co-authored-by: Jochen Haeussler <jochen.haeussler@heigit.org>

MichaelsJP committed:
feat(docker): Change the user example to 1000:1000

The user needs to be 1000:1000 if anything else than 0:0 wants to be chosen. Else the image needs to be rebuild with the correct gid and uid.

MichaelsJP committed:
feat(docker): Adjust the container config path

The path will now default to /home/ors/config allowing for easier mounts and a more streamlined way to configure the container.

MichaelsJP committed:
fix(docker): Migrate BASE_FOLDER to ORS_HOME

This removes the redundant BASE_FOLDER and replaces a couple of more hard coded paths to ORS_HOME.

MichaelsJP committed:
fix(docker): Remove redundant sanity check

MichaelsJP committed:
fix(docker): Set user id and gid to info

MichaelsJP committed:
fix(docker): Make the writeability and existens check explicit

It is good to have a more in depth feedback on what goes wrong.

MichaelsJP committed:
fix(docker): Explain the user option a bit clearer and add 1000:1000 to the checks

MichaelsJP committed:
fix(docker): Fix the indentation and add the individually explanation

MichaelsJP committed:
fix(docker): Migrate to the default jar location

This avoids users mounting and accidentally fiddling with the jar file.

MichaelsJP committed:
fix(docker): Add another even clearer explanation

MichaelsJP committed:
fix(docker): Fix port for healthcheck

jhaeu committed:
chore: added new docker directory to ignore files

jhaeu committed:
feat: changed volume name

MichaelsJP committed:
fix(workflow): Fix the wrong config volume mapping for the test workflow

jhaeu committed:
feat: evaluating ORS_TAG in docker compose

jhaeu committed:
feat: add docker compose file to release artifacts

MichaelsJP committed:
fix(docker-compose): Adjust the image tag to local/openrouteservice:latest

This ensures that users that checkout the compose from main will either fail or if it was cloned with the repo will have the latest repo stage build as their docker image. The image line is adjusted in the release action so that the release artifacts compose files will always point to their subsequent image version on docker hub.

MichaelsJP committed:
ci(release): Merge the new artifact release with the docker image artifact release

MichaelsJP committed:
fix(mvn): Add -q at some places to reduce verbosity of maven

koebi commented on Dockerfile:
```diff
@@ -1,98 +1,61 @@
-    chown -R ors ${BASE_FOLDER}/tomcat ${BASE_FOLDER}/ors-core/logs ${BASE_FOLDER}/ors-conf ${BASE_FOLDER}/ors-core/data ${BASE_FOLDER}/tomcat/logs
-
-WORKDIR ${BASE_FOLDER}
+RUN apk update && apk add --no-cache bash openssl yq jq curl && \
```
why do we install openssl and curl here?

MichaelsJP replied:
Good question regarding OpenSSL and curl. I think remnants from earlier dev steps. Will remove them.

koebi commented on docker-entrypoint.sh:
```diff
@@ -1,61 +1,366 @@
+echo "# Container sanity checks #"
+echo "###########################"
+info "Running container as user $(whoami) with id $(id -u) and group $(id -g)"
+if [[ $(id -u) -eq 0 ]] || [[ $(id -g) -eq 0 ]] ; then
+  debug "User and group are set to root with id 0 and group 0."
+elif [[ $(id -u) -eq 1000 ]] || [[ $(id -g) -eq 1000 ]] ; then
+  debug "User and group are set to 1000 and group 1000."
+else
```
Given the line above, this seems redundant?

MichaelsJP replied:
The goal here is more to react with an explicit debug message on both scenarios individually. Either 0 or 1000 have different implications for the runtime container.

koebi replied:
Can these implications be easily mentioned? If so, we might want to do that.
If not, that's fine by me.

MichaelsJP replied:
Nope. It depends on how they read the documentation and what way they set it up. When people temper with permissions, the container needs to feed back what it is ultimately using. Else it'll become a black box in that regard.

koebi commented on docker-entrypoint.sh:
```diff
@@ -1,61 +1,366 @@
-echo "ORS Path: ${ors_base}"
-echo "Catalina Path: ${catalina_base}"
+mkdir -p "${ORS_HOME}"/{files,logs,graphs,elevation_cache,config} || warning "Could not create ${ORS_HOME} and folders"
+mkdir -p "${ORS_HOME}"/{files,logs,graphs,elevation_cache,config} || warning "Could not create ${ORS_HOME} and folders"
```
`ORS_HOME` is created above, so this should only report errors on folders.
Also, why is this done twice?

MichaelsJP replied:
adjusted

koebi commented on docker-entrypoint.sh:
```diff
@@ -1,61 +1,366 @@
+# get ors_config_location
+ors_config_location=${ORS_CONFIG_LOCATION:-""}
+# Unset the ORS_CONFIG_LOCATION to not interfere with the spring-boot application
+unset ORS_CONFIG_LOCATION
```
This is only used in line 217, so we should move it there?

MichaelsJP replied:
I think it's better to have related functionality in one place. The initialization of the container env is done on that section.

koebi replied:
Yes, but since these variables are only used there and possibly even overwritten, I'd say the belong to that section, and shouldn't be declared way before, so no one remembers what they even were?

MichaelsJP replied:
Well, it's a matter of taste.  If you think it brings more value we can do that.
Fine for me :+1:

koebi commented on docker-entrypoint.sh:
```diff
@@ -1,61 +1,366 @@
+# get ors.engine.elevation.cache_path
+ors_engine_elevation_cache_path=$(env | grep "^ors\.engine\.elevation\.cache_path=" | awk -F '=' '{print $2}')
+# get ors.engine.source_file
+ors_engine_source_file=$(env | grep "^ors\.engine\.source_file=" | awk -F '=' '{print $2}')
```
These are only used in lines 234 onwards, so we should move it to there?

MichaelsJP replied:
same as above. i think the init of the container env should be done in one place completely.

MichaelsJP replied:
Well, it's a matter of taste.  If you think it brings more value we can do that.
Fine for me :+1:

koebi replied:
I think it makes the code clearer and easier to read, yes.

MichaelsJP replied:
btw. they're used in line 183 the first time. so i'll move them there

koebi commented on docker-entrypoint.sh:
```diff
@@ -1,61 +1,366 @@
+if [[ "${ors_config_location}" = *.yml ]] && [[ -f "${ors_config_location}" ]]; then
+  success "Using yml config: ${ors_config_location}"
+elif [[ "${ors_config_location}" = *.json ]] && [[ -f "${ors_config_location}" ]]; then
+  success "Using json config: ${ors_config_location}"
```
Here, we could note that we're using the values from the ENV?

MichaelsJP replied:
yep

MichaelsJP replied:
what about the following line after this if else construct?
```
info "Any ENV variables will have precedence over configuration variables from config files."
```

koebi replied:
That is valid information for the user, but these two lines are only printed if the value is read from an ENV. So I'm thinking along the lines of "Using yml config from ENV" or similar?

MichaelsJP replied:
I'll change it to the following:
```
success "Using yml config from ENV: ${ors_config_location}"
success "Using json config from ENV: ${ors_config_location}"
```

koebi commented on docker-entrypoint.sh:
```diff
@@ -1,61 +1,366 @@
+fi
+
+if [[ -z "${ors_engine_source_file}" ]] && [[ "${ors_config_location}" = *.json ]]; then
+  ors_engine_source_file=$(extract_config_info "${ors_config_location}" '.ors.services.routing.sources[0]')
```
We could refactor this to reflect the:
```
does it exist?
   either get from yml
   or get from json
```
structure

MichaelsJP replied:
@jhaeu please revisit this section. You've reworked this quite a lot.

jhaeu replied:
I think the block I changed was the previous one. 
Pushed a refactored code block respecting @koebi 's suggestion (hope I got it right)

koebi reviewed (changes requested):
I do have a few comments on Dockerfile/compose/entrypoint.
I'll look at the utils and workflows independently.

MichaelsJP committed:
chore: Remove dev dependencies

Curl and openssl are not needed for the final image.

MichaelsJP committed:
fix: Remove redundant mkdir and adjust the log messages

koebi commented on docker-compose.yml:
```diff
@@ -1,26 +1,84 @@
+    # Be careful with long graph builds, as the healthcheck will fail and show 'unhealthy' during this time.
+    # This is especially useful if you run your ORS container in a production environment.
+    #healthcheck:
+    #  test: curl --fail http://localhost:8082/ors/v2/health || exit 1
```
I just found that this is where curl is needed. I think we should switch to wget here.

MichaelsJP replied:
why wget?

MichaelsJP replied:
and yes. that was the reason. good catch

koebi replied:
wget should be available by default, so we don't need to install anything else

MichaelsJP replied:
gotya

jhaeu committed:
refactor: review finding: change to nested if blocks

MichaelsJP committed:
fix(dockerfile): Move the config env setting further down and adjust the config loading messages

The enhanced messages ensure that the user knows that a config is either read by ENV or as a fallback when existing in the config folder.

MichaelsJP committed:
fix(dockerfile): Add curl again

Curl is necessary for the health check in the docker-compose.yml

MichaelsJP committed:
fix(dockerfile): Replace curl with wget

Wget is provided by default and can also be used for the health check. One dependency less :).

MichaelsJP committed:
fix(dockerfile): Move the config parsing closer to its first usage

MichaelsJP commented:
@koebi Cannot answer on your comments directly, so I'll answer it here:

> I think it makes the code clearer and easier to read, yes.

Fixed. I moved them to its first usage in line 182.

> Can these implications be easily mentioned? If so, we might want to do that. If not, that's fine by me.

Not really. It depends on how they read the documentation and what way they're configuring the container user. 
Based on that, a simple feedback from the image on the internal is useful. Else, the container will become a small black box.
